Map a Handle to the object path, not the old "handle" path, avoiding double redirection.

mwoodiupui opened the pull request:
## Description
The Handle resolver plugin returns URLs in the old "https://dspace.example.com/handle/123/456" form.  The Global Handle Registry fetches this from DSpace and sends it to the user's browser as a redirect.  The browser requests that URL, and DSpace sends a second redirect to the actual object path.  That causes excess traffic and delays.  This patch reworks the handle resolution code to return the object path URL directly, eliminating the second redirection.

This patch introduces a new table of object type "plurals".  That is part of introducing an unfortunate indirect coupling between dspace-api and dspace-server-webapp.  The form of our REST URLs should be encapsulated in dspace-server-webapp, but the handle code is (properly) in dspace-api.  I chose to duplicate this knowledge rather than introduce a circular dependency.

Also, a related `package.html` was incorrect since we went to Services, so I took the opportunity to replace it with a modern `package-info.java`.

## Instructions for Reviewers
List of changes in this PR:
* Cause the handle resolver to return `type/UUID` URLs rather than `handle/prefix/suffix` URLs.
* New test method to demonstrate correct operation.

Alan Orth has kindly provided a test procedure below.

## Checklist
- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

mwoodiupui committed:
Map a Handle to the object path, not the old "handle" path, avoiding a double redirection.

This introduces a new table of object type "plurals".

Also, a related 'package.html' was incorrect since we went to Services, so
I took the opportunity to replace it with a modern 'package-info.java'.

mwoodiupui committed:
Fix silly checkstyle error.

mwoodiupui committed:
Reload configuration files before testing.

Another test suite somehow interfered with testGetHandle and
testGetSiteHandle when it changed the 'handle.prefix' value.  The Site
object is created in the database very early in testing and had a
handle already assigned when this suite started.

alanorth commented:
Thanks @mwoodiupui. I ported this patch to DSpace 7.6 and tested resolving a handle using the built-in HTTP resolver.

# Before:

```console
# curl x.x.x.x:8000/10568/3
<html><head><title>Handle Redirect</title></head>
<body><a href="https://dspace.example.org/handle/10568/3">https://dspace.example.org/handle/10568/3</a></body></html>
```

# After:

```console
# curl x.x.x.x:8000/10568/3
<html><head><title>Handle Redirect</title></head>
<body><a href="https://dspace.example.org/collections/8ea4b611-1f59-4d4e-b78d-a9921a72cfe7">https://dspace.example.org/collections/8ea4b611-1f59-4d4e-b78d-a9921a72cfe7</a></body></html>
```

I think having one less redirect is a good thing. I can't comment on the strategy used to achieve it in the patch.

alanorth commented:
Another data point for this: we've been having problems with Altmetric not tracking our content since we upgraded to DSpace 7. After discussing with them we realized their bot does not execute JavaScript so never receives the final redirect to the Handle's UUID from Angular's routing.

tdonohue commented:
Discussed in today's DevMtg & there was general agreement that this was the correct direction.  @abollini offered to help here, and mentioned we may want to also look at https://github.com/DSpace/Remote-Handle-Resolver to see if it requires updates as well.

mwoodiupui commented:
This breaks a number of tests in dspace-server-webapp, and I have not found where the old-style expectations are generated.  Guidance would be appreciated.

misilot commented:
> Discussed in today's DevMtg & there was general agreement that this was the correct direction. @abollini offered to help here, and mentioned we may want to also look at https://github.com/DSpace/Remote-Handle-Resolver to see if it requires updates as well.

I think it would need to be. An example for us is `https://krex.k-state.edu/server/resolve/2097/1522` returns
```json
["https://krex.k-state.edu/handle/2097/1522"]
```

mwoodiupui committed:
Fix broken tests.

mwoodiupui committed:
Fix more tests.

mwoodiupui commented:
After much 'grep'ing and single-stepping I think I have a slightly better understanding of the control flow in Spring-Web.  I was able to find the sources of the remaining test failures.

mwoodiupui committed:
Add a rudimentary test suite for the Handle server plugin.

This only tests what I wanted to know for the present patch:  will the
Handle server receive (and return?) the intended redirect URL?

mwoodiupui commented:
@misilot As you can see in the patch, it was necessary to fix `HdlResolverRestControllerIT` to account for the new URL pattern, so fixing the plugin seems to have fixed `HdlResolverRestController`.

mwoodiupui commented:
A version for DSpace 7.6 is in production here.

alanorth commented:
> A version for DSpace 7.6 is in production here.

Running in production on DSpace 7.6 for a week now here as well.

misilot commented:
I tried applying to 7.6 but get the following when trying to build

```
 #11 98.28 [ERROR] /app/dspace-api/src/test/java/org/dspace/handle/DummyHandleDAO.java:[97,4] error: method does not override or implement a method from a supertype
```

alanorth commented:
Hey @misilot. Which patch did you use? I manually ported it in my branch of 7.6.3-SNAPSHOT as of a week or two ago and it's working: https://github.com/ilri/DSpace/commit/374069565a9a963c00d21e0de1437eb0dc91b8a4

misilot commented:
https://github.com/DSpace/DSpace/pull/9864.patch

alanorth commented:
The patch in this PR is against `main` so there has been some divergence from 7.6 I guess.

alanorth commented:
I've found a side effect of this patch: collection subscriptions now display the object paths instead of legacy handle paths.

Before:

```
This email is sent from DSpace based on the chosen subscription preferences.

Collection Subscriptions:
-------------------------
List of new/changed items : 

Building the future One Health workforce in eastern and southern Africa: Gaps and opportunities in higher education https://cgspace.cgiar.org/handle/10568/151724
```

After:

```
This email is sent from DSpace based on the chosen subscription preferences.

Collection Subscriptions:
-------------------------
List of new/changed items : 

Participatory rangeland management: Understanding women's engagement and implications for social change https://cgspace.cgiar.org/items/2d7241da-f64a-4eae-80ea-c68fd98fde93
```

As an aside, I was surprised to see this because collection subscription emails in DSpace 4/5/6 used canonical Handles! In any case, this needs to be considered.

alanorth commented:
Another side effect of this patch is that it breaks the DSONameService, used to find the names of DSpace objects, for example here on the site statistics page:

![Screenshot 2025-02-12 at 11-59-37 CGSpace Statistics](https://github.com/user-attachments/assets/9e32145b-2515-4add-b46d-b2952fdc6ee0)

mwoodiupui commented:
@alanorth Regarding the subscription emails:  is there a problem or only a surprise?  What *should* we be showing there?

mwoodiupui commented:
Note:  DSONameService is in the front-end:  `src/app/core/breadcrumbs/dso-name.service.ts`.

alanorth commented:
> @alanorth Regarding the subscription emails: is there a problem or only a surprise? What should we be showing there?

Hey @mwoodiupui, historically we showed the Handles here and this is preferred because those are persistent and we encourage users to use those instead of the repository URLs. Handles are "forever"!

> Note: DSONameService is in the front-end: src/app/core/breadcrumbs/dso-name.service.ts.

If this patch is merged then the frontend will need to be adapted. I could file an issue to track it if we want to pursue this PR...

mwoodiupui commented:
Regarding the emails:  so both the old form and the new are wrong.  We should give the real `hdl.handle.net` URL.

alanorth commented:
Hi @mwoodiupui, according to my notes from #9994 the DSpace 6 behavior was to show the canonical hdl.handle.net form:

```
New items are available in the collections you have subscribed to:

New Items: ILRI presentations: 5

      Title: One Health in Schools: Extending the One Health Approach to Safety Around Animals
    Authors: Richards, Shauna
    Type: Presentation
         ID: https://hdl.handle.net/10568/134818

      Title: Introducing COHESA Work on One Health in Schools
    Authors: Mutua, Florence K.
    Type: Presentation
         ID: https://hdl.handle.net/10568/134821
```

mwoodiupui committed:
Make subscription emails show Handle URIs, not DSpace Handle-based URIs.
Add test suite to prove this.

mwoodiupui commented:
Maybe we need a ticket for restoring the details and labels in the subscription email?  Looks like quite a bit was lost here in implementing Entities.
